csvgen/gen.go

```diff
@@ -0,0 +1,208 @@
+package csvgen
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"sort"
+
+	"github.com/pkg/errors"
+)
+
+type Main struct {
+	Filename string
+	RunsMin  int
+	RunsMax  int
+	NMin     int
+	NMax     int
+	Rows     uint64
+	Cols     uint64
+
+	Stdin  io.Reader `json:"-"`
+	Stdout io.Writer `json:"-"`
+	Stderr io.Writer `json:"-"`
+}
+
+func (m *Main) Run() error {
+	f, err := os.Create(m.Filename)
+	if err != nil {
+		return errors.Wrap(err, "opening output file")
+	}
+	defer f.Close()
+
+	bw := bufio.NewWriter(f)
+	defer bw.Flush()
+
+	iter := &Iterator{
+		RunsMax: m.RunsMax,
+		RunsMin: m.RunsMin,
+		NMin:    m.NMin,
+		NMax:    m.NMax,
+		Rows:    m.Rows,
+		Cols:    m.Cols,
+	}
+
+	iter.Generate(bw)
+	return nil
+}
+
+type Iterator struct {
+	RunsMin int
+	RunsMax int
+	NMin    int
+	NMax    int
+	Rows    uint64
+	Cols    uint64
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func (iter *Iterator) Generate(w io.Writer) {
+	contsPerRow := iter.Cols / 65536
+	if iter.Cols%65536 > 0 {
+		contsPerRow += 1
+	}
+
+	var r uint64
+	for r = 0; r < iter.Rows; r++ {
+		var c uint64
+		for c = 0; c < contsPerRow; c++ {
+			numRuns := rand.Intn(iter.RunsMax-iter.RunsMin) + iter.RunsMin
+			if numRuns > iter.NMax {
+				numRuns = iter.NMax - 1
+			}
+			numN := rand.Intn(iter.NMax-iter.NMin) + iter.NMin
+			if numN < numRuns {
+				if numRuns > iter.NMax {
+					numRuns = numN
+				} else {
+					numN = numRuns
+				}
+			}
+			fmt.Fprintf(os.Stderr, "gen container: N: %d, Runs: %d\n", numN, numRuns)
+			cont := randomContainer(numN, numRuns)
+			for _, col := range cont.Bits() {
+				fmt.Fprintf(w, "%v,%v\n", r, col)
```

Comment:
it seems like `col` here needs to include the container offset. something like 
```fmt.Fprintf(w, "%v,%v\n", r, (col + (c * 65536)))```
Without that, all the column values are in the range of 0-65535.